src/test/java/com/google/vinet/servlets/RegistrationServletTest.java

```diff
@@ -0,0 +1,374 @@
+    System.setErr(new PrintStream(errorStream));
+  }
+
+  @AfterEach
+  void resetErrorStream() {
+    System.setErr(System.err);
+  }
+
+  @BeforeAll
+  static void setUp() {
+    helper.setUp();
+  }
+
+  @AfterAll
+  static void tearDown() {
+    helper.tearDown();
+  }
+
+  /**
+   * Get any messages that were printed on the error stream
+   *
+   * @return A string containing all errors
+   */
+  private String getErrors() {
+    return errorStream.toString().trim();
+  }
+
+  /**
+   * Sets up the user service mock to have a logged-in and valid user
+   */
+  private void setupUser() {
+    when(userService.isUserLoggedIn()).thenReturn(true);
+    when(userService.getCurrentUser()).thenReturn(user);
+    when(user.getUserId()).thenReturn("anuserid");
+  }
+
+  /**
+   * Sets up an argument captor to get the HTTP status code from the response
+   *
+   * @return The argument captor that will contain the HTTp status code, as an Integer
+   */
+  private ArgumentCaptor<Integer> captureResponseStatus() {
+    ArgumentCaptor<Integer> valueCapture = ArgumentCaptor.forClass(Integer.class);
+    doNothing().when(response).setStatus(valueCapture.capture());
+    return valueCapture;
+  }
+
+  /**
+   * Sets the HTTP request parameters. Can also be used to pass null values.
+   */
+  private void setRequestParameters(
+          String firstname, String lastname, String type, String latitude, String longitude) {
+    Map<String, String[]> parameterMap = new HashMap<>();
+    parameterMap.put("firstname", new String[]{firstname});
+    parameterMap.put("lastname", new String[]{lastname});
+    parameterMap.put("type", new String[]{type});
+    parameterMap.put("latitude", new String[]{latitude});
+    parameterMap.put("longitude", new String[]{longitude});
+    when(request.getParameterMap()).thenReturn(parameterMap);
+  }
+
+  /**
+   * Sets HTTP request parameters, but only if each value is not null
+   */
+  private void setNonNullRequestParameters(
+          String firstname, String lastname, String type, String latitude, String longitude) {
+    Map<String, String[]> parameterMap = new HashMap<>();
+    if (firstname != null) parameterMap.put("firstname", new String[]{firstname});
+    if (lastname != null) parameterMap.put("lastname", new String[]{lastname});
+    if (type != null) parameterMap.put("type", new String[]{type});
+    if (latitude != null) parameterMap.put("latitude", new String[]{latitude});
+    if (longitude != null) parameterMap.put("longitude", new String[]{longitude});
+    when(request.getParameterMap()).thenReturn(parameterMap);
+  }
+
+  /**
+   * Asserts that the HTTP response to this request was a certain value
+   *
+   * @param code The HTTP response code that the response contains
+   */
+  private void assertResponseCode(int code) {
+    ArgumentCaptor<Integer> valueCapture = captureResponseStatus();
+    registrationServlet.doPost(request, response);
+    assertEquals(code, valueCapture.getValue());
+  }
```

Comment:
you can do this slightly easier using Mockito's `verify()`. See: https://github.com/googleinterns/step226-2020/blob/396cee07d6ae6e059116137d295574577625627a/src/test/java/com/google/vinet/servlets/LogoutServletTest.java#L93 The syntax is `verify(object).someMethod(someArgument)` which will verify that `object` had its `someMethod` method called with `someArgument`  as its argument. (`object` must be a mock of course for this to work)

Reply:
Very interesting. I have changed it to use verify() instead of an ArgumentCaptor.